### PR TITLE
Adds method to list nodes on a wire

### LINKED ISSUE #1502: Need a data structure to list the gates on qubits
### What is the expected behavior?
Every qubit wire is given a list stores the gates operated on it in order. It will help DAGCircuit manipulation.

### PR DESCRIPTION
**Pull request by maddy-tod**
### Summary
Fixes #1502 

Adds a method `nodes_on_wire()` to the `DAGCircuit`. This method yields all the nodes in the DAG which affect the given wire. The method takes 2 parameters - 
`wire` : the wire to be looked at
`only_ops`(optional) : default to `False`, if `True` then only the op nodes are returned, otherwise all nodes on the wire are returned.

This will be affected by #1815 and will be updated to return DAGNodes instead.

### CODE REVIEW

**Comment by diego-plan9** [comment]
@maddy-tod while you are working on the file, can you rename `qiskit/dagcircuit/_dagcircuit.py` to `qiskit/dagcircuit/dagcircuit.py` (ie. removing the underscore) at a convenient time (for example, as the last commit in this PR)? Since you are one of most active contributors working on the file, it would make it smoother if you took care of it during one of your PRs, to avoid some conflicts.

File: qiskit/dagcircuit/_dagcircuit.py [hunk 1]
```diff
@@ -1371,6 +1371,56 @@ def collect_runs(self, namelist):
...
+        Yield:
+             node: the successive ops on the given wire
+        Raises:
+            DAGCircuitError: if the given wire doesn't exist in the DAG
+        """
+
+        starting_node = None
+
+        # Find the starting node for this wire
+        for node_id in self.input_map.values():
```
[the remark this answers is not in the document]
Comment (maddy-tod) [reply qiskit/dagcircuit/_dagcircuit.py]: No difference! Updated

**Comment by maddy-tod** [comment]
I have also renamed `qiskit/dagcircuit/_dagnode.py` to `qiskit/dagcircuit/dagnode.py` after #1815 .

**Comment by ajavadia** [comment]
This looks good, there were some conflicts that I resolved. 

However during the review, I noticed that the `DAGNode.__eq__` method is defined so that it can compare `DAGNode` against `int`. This was probably done due to some convenience, but I think it is a bit of an unexpected behavior to compare objects of different types. It took me a while to understand why the test was comparing against ints, when the `nodes_on_wire()` method is returning DAGNodes. Can we fix this in a follow up PR?

**Comment by maddy-tod** [comment]
@ajavadia That was done to allow for easier testing (there was a comment in the code about it at some stage but it must have been removed). I think this was mainly because the older version of `DAGNode.__eq__` only allowed for identical `DAGNode` objects to compare to equal. This meant that we couldn't test that, for example, a method returned the nodes in the right order without looking at the `node_id`, as if we simply made identical objects to what we were expecting they wouldn't compare as equal. It was neatest to put it in `DAGNode.__eq__` as otherwise all the testing would have to access the private variable `_node_id`.

The work @1ucian0 is doing should fix this but we could also refactor it out by having a method to return the  `_node_id`/compare a given val with the `_node_id` to see if it is equal.

**Pull request merged**